### ISSUE TITLE
Best practice to set an Entity's relationship

### ISSUE THREAD

**Issue by p-nicolaou** [ffried/FFCoreData#23]
Hello @ffried 

Great work here! I have a question for you. If my Entity A has a relationship to another Entity B, what's the best way of updating that property inside Entity A's update(:)? As update(:) doesn't hold any reference to NSManagedObjectContext, I can't call Entity's B findFirst(:).

**Comment by ffried** [comment]
You actually have multiple options here 🙂 

First of all, inside the `update(from:)` method, you should have access to the `managedObjectContext` property of the instance of Entity A.
Depending on how you decode your objects, you might also have access to `NSManagedObjectContext.decodingContext(at:)`. If you're using the convenience extensions on `JSONDecoder`/`PropertyListDecoder`, `.decodingContext(at:)` will work. Otherwise, you need to make sure, that your decoding call happens inside a `asDecodingContext(do:)` call on the managed object context. Otherwise `NSManagedObjectContext.decodingContext(at:)` will throw an error that no context has been set up as decoding context.

**Comment by p-nicolaou** [comment]
Hello

Thank you for your quick response :)

I will test this approach out and let you know. Can I ask you though, is it safe to access the reference of `NSManagedObjectContext` with both ways in the same class? What I mean by that is, imagine I have a **CoreDataClass** that implements both public static `func findOrCreate(for dto: DTO, in context: NSManagedObjectContext) throws -> Self` and `public func update(from dto: DTO) throws`. Now, is it safe to use `context` in the first method but inside `update(:)` to use `.decodingContext(at:)`? (obviously after I have called `asDecodingContext(do:)`. In theory if it's the same object, I won't have any problems.

**Comment by ffried** [comment]
What `asDecodingContext(do:)` does is to store the context on which this method is called as the decoding context for the current thread. `.decodingContext(at:)` then simply gets the stored context for the current thread. The `codingPath` parameter is only used for the error when no context is found to make it easier to find where the issue occurred.

Back to your question: It slightly depends on from where the `update(from:)` call is originating from. In the currently released version (5.0.0), if it originates from the convenience implementation of `findOrCreate(for:in:)` and this convenience implementation wasn't called from within `asDecodingContext(do:)`, it's **not** safe to access `.decodingContext(at:)`. However, this should actually be the case, which is why I've just pushed a small fix for this.
Since you're implementing `findOrCreate(for:in:)` yourself, you'd be safe to use `.decodingContext(at:)` inside `update(from:)` as long as you perform the latter call inside `.asDecodingContext(do:)` like you described.
And yes, in this case `.decodingContext(at:)` will return the same object as `context`.

**Comment by ffried** [comment]
Btw., in my use cases, I usually do the following inside `update(from:)` to get the context:
```swift
func update(from dto: DTO) throws {
    let context = try managedObjectContext ?? .decodingContext()
    ...
}
```

The reason `update(from:)` has no `context` parameter isn't, that I wouldn't have access to one at all call sites. It's more to avoid having to make the following assertion:
```swift
func update(from dto: DTO, in context: NSManagedObjectContext) throws {
    assert(context == managedObjectContext)
}
```
You could really even do `guard let context = managedObjectContext else { return }` as first line in your `update(from:)` method, since if your object has no context, it very likely isn't valid anymore.
